### DISCUSSION TITLE
Understanding the design choices behind example usage of SpanLinks (as per documentation)

### DISCUSSION THREAD

**Discussion by Ayush-Kaushik** [open-telemetry/opentelemetry.io#5632] [Information Architecture]
Documentation Link: https://opentelemetry.io/docs/concepts/signals/traces/#span-links

## Background:

I have a lambda `ProcessEvents` which processes incoming events in batches. Each event contains unique propagated context which comes from another microservice (lets call it `SendEvents`)

Now when `ProcessEvents` is invoked, we have a parent span generated. Inside the ProcessEvents lambda, we loop through the batch of events and process them one by one (this operation is denoted by its own unique span created within this loop). This is also where we extract the propagated context from the incoming event.

The scenario mentioned above is a perfect use case for SpanLinks. According to OpenTelemetry documentation, the best approach to show the association among propagated context from SendEvents, Parent span of ProcessEvents and child span created when each event is processed in loop is:

```
The child span should be associated with `ProcessEvents` parent span and the propagated context should be added as a SpanLink.
```

but why was this approach chosen in first place? Why didn't we go with this approach instead:

```
Should we add `ProcessEvents` span as a SpanLink for the child span while making the PropagatedContext as the parent.
```

Does one approach hold merit over the other? It'll be helpful if some more clarification is provided over SpanLinks usage in documentation.

**Comment by trask** [comment]
hi @Ayush-Kaushik! I'm going to transfer this to the website repo, since it looks like the source for [this page](https://opentelemetry.io/docs/concepts/signals/traces/#span-links) lives [over there](https://github.com/open-telemetry/opentelemetry.io/blob/48d10d612508b6096d9c62250f6ed5b212271ae3/content/en/docs/concepts/signals/traces.md?plain=1#L295)

separately, you may find this helpful: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md#trace-structure

**Comment by svrnm** [comment]
>  According to OpenTelemetry documentation, the best approach ...

the page you linked does not provide a suggested best approach, it just gives an example.

**Reply by pg94au** [reply]
The additional link that was provided by Trask does include the following:

"Exclusively for single messages scenarios, the "Process" span MAY use the message's creation context as its parent, thus achieving a direct parent-child relationship between producer and consumer(s)."

In this case, is that disqualifying batch message processing, or would it be appropriate to consider each message received in a batch as that _single_ message?

Is this statement referring to a scenario where an activity is taking place as a direct consequence of more than one initiating message, where it is impossible to identify only a single originator as the one parent?

**Reply by trask** [reply]
hi @pg94au, I'd suggest asking these more detailed questions over in the [semantic-conventions repo](https://github.com/open-telemetry/semantic-conventions)
